SmallPermutationGroups

dwbump opened the pull request:
<!-- ^^^^^
Implement GAP SmallGroups as Sage Permutation Grops
-->
### 📚 Description

GAP has a database of Small Groups identified by a pair of numbers [n,k] where n is the order and k is an integer used to differentiate different groups of the same order. Thus k is bounded by oeis(1)[n], the number of isomorphism classes of groups of order n. Since Sage PermutationGroups have many useful methods, it is good to have PermutationGroup realizations of GAP SmallGroups. The docstring verifies that the new SmallPermutationGroup class is able to construct all groups of order <= 64, though of course groups of order >64 are also available.

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [ ] I have linked an issue or discussion.
- [ x] I have created tests covering the changes.
- [ x] I have updated the documentation accordingly.

### ⌛ Dependencies
<!-- List all open pull requests that this PR logically depends on -->
<!--
- #xyz: short description why this is a dependency
- #abc: ...
-->

dwbmscz committed:
SmallPermutationGroups

dwbmscz committed:
SmallPermutationGroups

dimpase reviewed (commented):
Can you use libgap rather than gap? I.e. 
```
self._gap_small_group = libgap.SmallGroup(order,gap_id)
```

dimpase commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -87,7 +89,9 @@
 from pathlib import Path
 
 from sage.rings.all import Integer
+from sage.interfaces.gap import gap
```
please no more dependencies on pexpect gap. We are trying to get rid of it in #26902

dwbump replied:
If we do the suggested self._gap_small_group = libgap.SmallGroup(order,gap_id) then we lose some functionality, needed for an application (FusionDouble, probably an upcoming PR). We need the methods conjugacy_classes_representatives and centralizer.

I haven't figured out why these do not work when we use libgap. I tried making SmallPermutationGroup inherit from PermutationGroup_generic since that class has the needed methods.

dimpase replied:
is `FusionDouble` a new functionality, not in Sage yet?

dwbump replied:
Yes, analogous to FusionRing one can construct the Fusion Rings of the Drinfeld Double of a finite group.

dwbmscz committed:
added doctest that is broken if we change to libgap

dwbmscz committed:
added doctest that is broken if we change to libgap

dimpase committed:
safely compute Image

as explained by @ChrisJefferson:

Image(IsomorphismPermGroup(g)) doesn't return what you want (always) -- as it can return a larger group (this behaviour may be changed in a future version of GAP to what you want, but it isn't that right now -- gap-system/gap#5355 ).

Instead you want to call Image(IsomorphismPermGroup(g),g)

dwbmscz committed:
Remark on generators and degree in SmallPermutationGroup docstring

dwbmscz committed:
pass correct parameters to libgap group __init__

dimpase committed:
no need to pass gens

tscrim commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -3435,3 +3438,81 @@ def codegrees(self):
+
+class SmallPermutationGroup(PermutationGroup_generic):
+    def __init__(self, order, gap_id):
+        r"""
```
I would move this doc to the class level and have a `TestSuite` doctest in `__init__`.

dwbump replied:
> That’s likely because there are some features missing that are expected. I will take a look tonight.

Thanks!

tscrim commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -3435,3 +3438,81 @@ def codegrees(self):
+        self._order = order
+        self._gap_id = gap_id
+        self._gap_small_group = libgap.SmallGroup(order,gap_id)
+        self._gap_permutation_group = self._gap_small_group.IsomorphismPermGroup().Image(self._gap_small_group)
```
Why do we need this attribute? It already exists in the base class (under a different name).

tscrim commented on src/sage/groups/perm_gps/permgroup.py:
```diff
@@ -247,6 +247,8 @@ def from_gap_list(G, src):
     # src is a list of strings, each of which is a permutation of
     # integers in cycle notation. It may contain \n and spaces.
+    if G is None:
+        return [eval(("[%s]"%g).replace(" ","").replace("\n","").replace(")(","),(")) for g in src]
```
This seems to no longer be used. Can we revert this change?

tscrim commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -3435,3 +3438,81 @@ def codegrees(self):
+        Gap realizes some small groups as PermutationGroup, others as PcGroups
+        (polycyclic groups). The :class:`SmallPermutationGroup` class always
+        returns a PermutationGroup, but in the process of creating this group
+        a gap SmallGroup is generated. This method returns that group.
```
```suggestion
        Return the GAP small group object corresponding to ``self``.
        
        Gap realizes some small groups as PermutationGroup, others as PcGroups
        (polycyclic groups). The :class:`SmallPermutationGroup` class always
        returns a PermutationGroup, but in the process of creating this group
        a gap SmallGroup is generated. This method returns that group.
```

dwbump replied:
I made most of the changes requested by Travis, except one comment above that I am asking for clarification on.

Unfortunately if I add a testsuite to __init__ (currently commented out) there are many failures.

tscrim replied:
I meant to put my comment below here.

You also have underindented the `EXAMPES::` block I think.

dwbump replied:
> You also have underindented the EXAMPES:: block I think.

Fixed.

tscrim replied:
The problem comes from the fact the permutation group expects that `_order` is a method. There are 2 ways to fix this. The first would be to change `_order` to, e.g., `_group_order` and override the `_order()` method to simply return `_group_order`. The second would be to override `order()` to simply return `self._order`. The method `_order()` only seems to be used in the `order()` method. So I would go with option 2.

tscrim replied:
You will also need to reset the `cardinality = order` alias.

tscrim replied:
So I added this at the end of the class and the `TestSuite` passed:
```python

    def order(self):
        return self._order

    cardinality = order
```

dwbump replied:
> So I added this at the end of the class and the `TestSuite` passed:

I fixed this alternatively by just renaming the `_order` attribute `_n` and the testsuite passes.

>
> ```python
>     def order(self):
>         return self._order
> 
>     cardinality = order
> ```

dwbmscz committed:
Changes requested by Travis Scrimshaw

dwbmscz committed:
fix underindented block

dwbmscz committed:
SmallPermutationGroup attribute must not be called _order

dwbmscz committed:
make gap_permutation_group a local variable

dimpase commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -88,6 +90,7 @@
 
 from sage.rings.all import Integer
 from sage.libs.gap.libgap import libgap
+from sage.groups.perm_gps.permgroup import from_gap_list
```
This import is not needed any more.

dimpase committed:
remove unneeded import

fingolfin commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -3435,3 +3437,91 @@ def codegrees(self):
+    r"""
+    A GAP SmallGroup, returned as a permutation group.
+
+    GAP contains a database of groups, each identified by its GAP id.
```
GAP contains many databases of groups, with an id. This should clarify which one you are referring to: the small groups library? transitive groups? primitive groups? perfect groups? ...

if it is the small groups library (SGL), the correct name would be more like "small groups id" -- note that Magma uses the exact same ids

fingolfin commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -3435,3 +3437,91 @@ def codegrees(self):
+            sage: G = SmallPermutationGroup(12,4); G
+            Group of order 12 and GAP Id 4 as a permutation group
+        """
+        return "Group of order %s and GAP Id %s as a permutation group"%(self._n, self._gap_id)
```
Why "Id" and not "id" ?

dwbump replied:
I revised the docstrings following these comments and pushed. Thanks!

fingolfin replied:
It still says `Id` instead of `id` in several places

fingolfin commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -3435,3 +3437,91 @@ def codegrees(self):
+        r"""
+        Return the GAP small group object corresponding to ``self``.
+
+        Gap realizes some small groups as PermutationGroup, others as PcGroups
```
```suggestion
        GAP realizes some small groups as PermutationGroup, others as PcGroups
```
But actually, these groups are provided by the Small Groups Library, a package for GAP, not by the GAP core itself. Just mentioning in case you are for precise attribution...

fingolfin replied:
You could also make it more precise: all solvable groups are PcGroups, the non-solvable ones permutation groups

fingolfin commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -3435,3 +3437,91 @@ def codegrees(self):
+        Gap realizes some small groups as PermutationGroup, others as PcGroups
+        (polycyclic groups). The :class:`SmallPermutationGroup` class always
+        returns a PermutationGroup, but in the process of creating this group
+        a gap SmallGroup is generated. This method returns that group.
```
```suggestion
        a GAP SmallGroup is generated. This method returns that group.
```

dwbump committed:
Update src/sage/groups/perm_gps/permgroup_named.py

Docstring revision in SmallPermutationGroup: Gap -> GAP

Co-authored-by: Max Horn <max@quendi.de>

dwbump committed:
Update src/sage/groups/perm_gps/permgroup_named.py

Docstring revision in gap_small_group method of SmallPermutationGroup: gap -> GAP

Co-authored-by: Max Horn <max@quendi.de>

dwbmscz committed:
Docstring revisions in SmallPermutationGroup recommended by fingolfin.

dimpase committed:
Merge branch 'develop' into small-permutation-groups

dwbmscz committed:
SmallPermutationGroup returns self._n as its order, to avoid recomputing it.

dwbmscz committed:
Merge branch 'small-permutation-groups' of github.com:dwbump/sage into small-permutation-groups

tscrim commented on src/sage/groups/perm_gps/permgroup_named.py:
```diff
@@ -3435,3 +3437,104 @@ def codegrees(self):
+        [ 1  1 -1  1 -1 -1]
+        [ 2  0 -2 -1  0  1]
+        [ 2  0  2 -1  0 -1]
+        sage: all(SmallPermutationGroup(n,k).id()==[n,k] for n in [1..64] for k in [1..oeis(1)[n]])
```
This test needs to be marked with `# optional - internet` and/or replaced by something that is manually inputted. (I would at least include part of the latter since it means it will get more regularly tested.)

fingolfin replied:
Why does it query data from OEIS that is part of the very small groups library that is  eing wrapped here in the first place?

dwbump replied:
I replaced the oeis(1) by a call to libgap.NumberSmallGroups.

dwbmscz committed:
replace oeis(1) by call to libgap.NumberSmallGroups

dwbmscz committed:
whitespace in def numgps (permgroup_named.py)

tscrim reviewed (approved):
Yes, that's right. Thank you.
